Add metrics and events for outgoing postgresql connections

thejoelw opened the pull request:
Fixes #85031

It's relatively simple to track total online connections, but to track the subset of connections in-use (currently executing a query), it's a little trickier. To accomplish this there's a change to the Connection interface: instead of exposing the underlying connection as a reference, expose a thin Lease class that lives on the stack and reflects the usage of the connection (typically by a query). Its constructor and destructor respectively increment and decrement the in-use counter (PostgreSQLClientConnectionsInUse).

<img width="860" height="469" alt="Screenshot 2025-08-15 at 2 03 47 AM" src="https://github.com/user-attachments/assets/7487ee16-3a5f-47ac-87c6-8055ac1e6122" />

<!---
A technical comment, you are free to remove or leave it as it is when PR is created
The following categories are used in the next scripts, update them accordingly
utils/changelog/changelog.py
tests/ci/cancel_and_rerun_workflow_lambda/app.py
-->
### Changelog category (leave one):
- New Feature

### Changelog entry:
Users can now see metrics and events describing outgoing postgresql connections.

### Documentation entry for user-facing changes

- [X] Documentation is written (mandatory for new features)

Added docs to docs/en/operations/system-tables/metrics.md

thejoelw committed:
Added connection counting metrics and profile events to postgres::Connection.
To track the subset of connections in-use, there's a change to the Connection interface: instead of exposing the underlying connection as a reference, expose a thin Lease class that is meant to live on the stack and reflect the usage of the connection. Its constructor and destructor respectively increment and decrement the postgres in-use counter.
This commit also contains updates to use the Lease interface.

thejoelw committed:
Merge remote-tracking branch 'upstream/master' into outgoing_postgresql_metrics

thejoelw committed:
Added docs for outgoing postgresql metrics

thejoelw committed:
Added an integration test for outgoing postgres metrics and events

thejoelw committed:
Fix linting errors

vdimir commented on src/Core/PostgreSQL/Connection.cpp:
```diff
@@ -2,11 +2,46 @@
+    CurrentMetrics::add(CurrentMetrics::PostgreSQLClientConnectionsIdle, 1);
+}
+
+Connection::Lease::Lease(const Lease & other) : Lease(*other.connection) {}
```
On copy we update metrics again for same connection, do we really need to do so? And `operator=` below does not. Do we really want to copy `Lease`, can't we make it noncopyable or move-only?

vdimir commented on docs/en/operations/system-tables/metrics.md:
```diff
@@ -574,7 +574,19 @@ Number of asynchronous inserts that are waiting for flush.
-Number of client connections using PostgreSQL protocol
+Number of incoming connections using PostgreSQL protocol
+
+### PostgreSQLClientConnections {#postgresqlconnection}
+
+Number of outgoing PostgreSQL client connections
+
+### PostgreSQLClientConnectionsIdle {#postgresqlconnection}
+
+Number of outgoing PostgreSQL client connections with no active queries
+
+### PostgreSQLClientConnectionsInUse {#postgresqlconnection}
```
```suggestion
### PostgreSQLClientConnections {#postgresqlclientconnections}

Number of outgoing PostgreSQL client connections

### PostgreSQLClientConnectionsIdle {#postgresqlclientconnectionsidle}

Number of outgoing PostgreSQL client connections with no active queries

### PostgreSQLClientConnectionsInUse {#postgresqlclientconnectionsinuse}
```

thejoelw replied:
Done

vdimir commented on src/Core/PostgreSQL/Connection.h:
```diff
@@ -33,14 +33,31 @@ struct ConnectionInfo
     void execWithRetry(const std::function<void(pqxx::nontransaction &)> & exec);
 
-    pqxx::connection & getRef();
+    Lease getLease();
```
```suggestion
    [[nodiscard]] Lease getLease();
```

thejoelw replied:
Done

vdimir commented on src/Core/PostgreSQL/ConnectionHolder.h:
```diff
@@ -40,19 +40,21 @@ class ConnectionHolder
     }
 
-    pqxx::connection & get()
+    postgres::Connection::Lease getLease()
```
```suggestion
    [[nodiscard]] postgres::Connection::Lease getLease()
```

vdimir commented on src/Core/PostgreSQL/ConnectionHolder.h:
```diff
@@ -40,19 +40,21 @@ class ConnectionHolder
             {
-                connection->getRef().reset();
+                // First try to close the connection gracefully.
+                connection->close();
             }
             catch (...)
             {
+                // If that fails for some reason, hard-reset it.
```
Reset calls `~Connection` with also does `Connection::close`, not not clear why should we do like this.

thejoelw replied:
I'm a little fuzzy on the expected logic here. The original code calls `pqxx::connection::reset()`, which is in the header but undocumented as far as I can see: https://libpqxx.readthedocs.io/stable/classpqxx_1_1connection.html#details

I've restored the behavior of calling it, but I suspect `close` or something similar might be better. I'm not sure the pros/cons.

Finally, when the connection is reset won't a null ptr be returned to the pool?

vdimir reviewed (commented):
Thanks for your contribution! Overall looks great, I've added some comments

thejoelw committed:
Address PR feedback

thejoelw committed:
Fix the integration test

thejoelw committed:
Merge remote-tracking branch 'upstream/master' into outgoing_postgresql_metrics

thejoelw committed:
Fix tests, since PG connections are auto-garbage-collected

thejoelw committed:
Removed assertions, since PG connection cleanup is unreliable

thejoelw committed:
Merge remote-tracking branch 'upstream/master' into outgoing_postgresql_metrics

grantholly-clickhouse committed:
Merge branch 'master' into outgoing_postgresql_metrics

grantholly-clickhouse commented:
@thejoelw I have brought in the upstream master branch and got CI running again.  If I messed anything up in the merge, let me know.  I'm hoping to get this PR finished and merged in.

grantholly-clickhouse commented on src/Core/PostgreSQL/Connection.cpp:
```diff
@@ -74,6 +125,25 @@ void Connection::connect()
+void Connection::resetConnection()
+{
+    // pqxx::connection::reset(), not std::unique_ptr::reset()
+    connection->reset();
```
```suggestion
    connection.get()->reset();
```

alexey-milovidov committed:
Delete aspell-dict.txt (removed in master)

`aspell-dict.txt` was deleted in master along with the `aspell` and
`codespell` style checks. Removing the file here so the PR can merge
cleanly without a modify/delete conflict.
